src/main/scala/cromwell/engine/workflow/WorkflowActor.scala

```diff
@@ -196,129 +239,156 @@ case class WorkflowActor(workflow: WorkflowDescriptor, backend: Backend)
-    case m =>
+    case Event(CallCompleted(callKey, outputs, returnCode), data) if callKey.isPersistedRunning(data) =>
+      callCompletedWhileAborting(callKey, outputs, returnCode)
+    case Event(CallCompleted(collectorKey: CollectorKey, outputs, returnCode), data) if !collectorKey.isPending(data) =>
+      // Collector keys don't have to be in Running state.
+      val updatedData = data.addPersisting(collectorKey)
+      callCompletedWhileAborting(collectorKey, outputs, returnCode)
+    case Event(CallAborted(callKey), data) if callKey.isPersistedRunning(data) =>
+      persistStatus(callKey, ExecutionStatus.Aborted, None)
+      val updatedData = data.addPersisting(callKey)
+      (if (isWorkflowAborted) transition(WorkflowAborted) else stay()) using updatedData
+    case Event(CallFailed(callKey, returnCode, failure), data) if callKey.isPersistedRunning(data) =>
+      persistStatus(callKey, ExecutionStatus.Failed, returnCode)
+      val updatedData = data.addPersisting(callKey)
+      (if (isWorkflowAborted) transition(WorkflowAborted) else stay()) using updatedData
+    case Event(m, _) =>
       logger.error("Unexpected message in Aborting state: " + m.getClass.getSimpleName)
-      if (isWorkflowAborted) goto(WorkflowAborted) using NoFailureMessage else stay()
+      if (isWorkflowAborted) transition(WorkflowAborted) else stay()
   }
 
-  when(WorkflowAborted)(FSM.NullFunction)
-
   whenUnhandled {
     case Event(AbortWorkflow, _) =>
       context.children foreach { _ ! CallActor.AbortCall }
-      goto(WorkflowAborting) using NoFailureMessage
+      transition(WorkflowAborting)
+    case Event(callMessage: CallMessage, _) =>
+      logger.debug(s"Dropping message for ineligible key: ${callMessage.callKey.tag}")
+      // Running and Aborting have explicit handlers for Running and no pending writes, so either we are not in those
+      // states or the specified key is no longer Running and/or has pending writes.
+      stay()
+    case Event(PersistenceCompleted(callKey), data) =>
+      logger.debug(s"Got PersistenceCompleted(${callKey.tag})")
+      val updatedData = data.removePersisting(callKey)
+      stay using updatedData
+    case Event(PerformTransition(toState), data) if data.pendingExecutions.isEmpty && !stateName.isTerminal =>
+      // stateName is the *current* state, from which there should not be transitions.
+      logger.info(s"transitioning from $stateName to $toState.")
+      goto(toState)
+    case Event(PerformTransition(toState), _) if stateName.isTerminal =>
+      // Drop the message.  This is a race condition with a retried PerformTransition message when
+      // the FSM has transitioned to a terminal state between the original and retried messages.
+      stay()
+    case Event(message @ PerformTransition(toState), _) =>
+      logger.debug(s"Rescheduling transition request to $toState due to pending persists")
+      context.system.scheduler.scheduleOnce(100 milliseconds, self, message)
```

Comment:
Small notes, that are not deal breakers, and may stay as is:

* Generally prefer if the returned `Cancellable` are at least attempted to be `.cancel()`ed when not needed, instead of flying as a dead letter.
* The three instances of `100 milliseconds` (or maybe even the whole `scheduleOnce` call?) could be put into a constant (or method if wrapping `scheduleOnce`).